cedar-policy-validator/src/schema.rs

```diff
@@ -475,48 +475,50 @@ impl ValidatorSchema {
         self.entity_types.iter()
     }
 
-    /// Get the validator entity equal to an EUID using the component for a head
-    /// var kind.
-    pub(crate) fn get_entity_eq<'a, H, K>(&self, var: H, euid: EntityUID) -> Option<K>
-    where
-        H: 'a + HeadVar<K>,
-        K: 'a,
-    {
-        var.get_euid_component(euid)
-    }
-
-    /// Get the validator entities that are in the descendants of an EUID using
-    /// the component for a head var kind.
-    pub(crate) fn get_entities_in<'a, H, K>(
+    /// Get all entity types in the schema which can be descendants of the given
+    /// entity type in the entity type graph.
+    pub(crate) fn get_entity_types_in<'a>(
         &'a self,
-        var: H,
-        euid: EntityUID,
-    ) -> impl Iterator<Item = K> + 'a
-    where
-        H: 'a + HeadVar<K>,
-        K: 'a + Clone,
-    {
-        var.get_descendants_if_present(self, euid.clone())
+        entity: &'a EntityUID,
+    ) -> impl Iterator<Item = &Name> + 'a {
+        let ety = match entity.entity_type() {
+            EntityType::Concrete(ety) => Some(ety),
+            EntityType::Unspecified => None,
+        };
+
+        ety.and_then(|ety| self.get_entity_type(ety))
             .into_iter()
-            .flatten()
-            .map(Clone::clone)
-            .chain(var.get_euid_component_if_present(self, euid))
+            .flat_map(|ety| ety.descendants.iter())
+            .chain(ety.filter(|ety| self.is_known_entity_type(ety)))
```

Comment:
does this assume that all entity types can be `in` themselves?

Reply:
Yes. This function computes entities types where an `in` expression can evaluate to `true`, which can always happen for `e in e` even if that's otherwise not valid.